### ISSUE TITLE
Please review

### ISSUE THREAD

**Issue by rkcudjoe** [hookengine/hubstaff-rails#1]
Hey there, @urkle, do you mind reviewing this. I'm particularly interested in what you have to say about handling a `NoMethodError` that occurs when an app reaches the 10 requests/hour limit during authentication.

```
  def authenticate_and_save_auth_token(email,password)
    client_user = User.find_by_email(current_user.email)
    HUBSTAFF_CLIENT.authenticate(@hubstaff_email,@hubstaff_password)
    client_user.hubstaff_auth_token = HUBSTAFF_CLIENT.auth_token
    client_user.save!
  end
```

**Comment by urkle** [comment]
Is that a nomethoderror in your code? or from our API?? if it's from our API, then that is a bug on our side that I need to fix.

**Comment by rkcudjoe** [comment]
@urkle, everything works fine if the API is used as expected, making less that 10 per hour.

```ruby
pry(main)> client = Hubstaff::Client.new("pHR18-G-9c05NoyBtji3a8A2KsFKOuZcSZK4gT5V9vc")
=> #<Hubstaff::Client:0x007fcaab91d958
 @app_token="pHR18-G-9c05NoyBtji3a8A2KsFKOuZcSZK4gT5V9vc",
 @auth_token=nil>
pry(main)> client.authenticate("rkcudjoe@hookengine.com","hookdev001")
=> "5WZ1SCto37HBhH-AR1jn0kC3FXROO4b39CREMSyt_1U"
```
But when you make the 11th within an hour, you get the nomethoderror because the auth_token is never returned as expected.

It's definitely not the Rails app; and I'm not sure whether it's the client.

But, the API has logic that checks the rate limit before making every request. So I'm guessing the API should return an object that the programmer can work EVEN IF they exceed their rate limit and never get the data they wanted.

Right now, that object is nil.

```ruby
pry(main)> 11.times { client.authenticate("rkcudjoe@hookengine.com", "hookdev001") }  
NoMethodError: undefined method `[]' for nil:NilClass
from /Library/Ruby/Gems/2.0.0/gems/hubstaff-ruby-0.0.1/lib/hubstaff/client.rb:26:in `authenticate'
```
Honestly, I'm not exactly sure how to fix this. But everything seems ok except when people reach their rate limit, which could mean that's something that has be fixed by you.

**Comment by urkle** [comment]
So the issue is your code..  You are not handling the 429 error code at all.    The response from a 429 (rate limit exceeded) is a JSON object of the form {"error": "message"}.  and your code currently assumes it will always succeed..  It should not do this, there should be code in your core URL fetching code that catches the 429 and raises and error appropriately.  (this goes for the PHP and the .NET code as well)

And do note that ANY of the endpoints can return 429 or other errors and should be handled accordingly.

**Comment by rkcudjoe** [comment]
@urkle, you're right. I was able to fix the client so that it returns a JSON object for every request. So now, the programmer has an object to work with for every request.

```ruby
[1] response = client.authenticate("rkcudjoe@hookengine.com","hookdev001")
=> #<Faraday::Response:0x007ff333bb39f0
[2] response.status
=>429
[3] pry(main)> if response.status == 200
[3] pry(main)*   puts response.body
[3] pry(main)* else  
[3] pry(main)*   puts response.body["error"]
[3] pry(main)* end 
=> You can't make any more requests
```
Now the programmer can handle every response object however they'd like, depending on the status. 

I'm about to change the code to handle errors correctly in the sample Rails app; did you get a chance to [watch the video walkthrough](https://youtu.be/-5fRjR_V5do)/ read the tutorials?

Is there anything I should keep in mind for the sample Rails app?

**Comment by urkle** [comment]
Common practice would be to raise appropriate errors for the various status codes with the data attached to the raised exception.  returning a big blob of data and letting the user figure it out doesn't provide much value in a library. as, why would I use the library over just using Net::HTTP directly?

**Comment by urkle** [comment]
and really we don't care for a rails sample app as that's not the use-case we are wanting to show off here..   It's tools to export/sync the data, not access it real-time.

**Comment by rkcudjoe** [comment]
@urkle I did what you suggested, parse the error messages too so that the user knows exactly what's wrong. And that's been done for the all the other versions too.

**Comment by urkle** [comment]
I'll take a gander through all of these later in the week and verify everything works as expected.

**Comment by rkcudjoe** [comment]
@urkle did you get a chance to review these?
